ddl/cluster.go

```diff
@@ -245,31 +254,100 @@ func GetFlashbackKeyRanges(sess sessionctx.Context, startKey kv.Key) ([]kv.KeyRa
 		})
 	}
 
-	for i, ranges := range keyRanges {
-		// startKey smaller than ranges.StartKey, ranges begin with [ranges.StartKey, ranges.EndKey)
-		if ranges.StartKey.Cmp(startKey) > 0 {
-			keyRanges = keyRanges[i:]
+	return keyRanges, nil
+}
+
+func sendPrepareFlashbackToVersionRPC(
+	ctx context.Context,
+	s tikv.Storage,
+	r tikvstore.KeyRange,
+) (rangetask.TaskStat, error) {
+	startKey, rangeEndKey := r.StartKey, r.EndKey
+	var taskStat rangetask.TaskStat
+	bo := tikv.NewBackoffer(ctx, flashbackMaxBackoff)
+	for {
+		select {
+		case <-ctx.Done():
+			return taskStat, errors.WithStack(ctx.Err())
+		default:
+		}
+
+		if len(rangeEndKey) > 0 && bytes.Compare(startKey, rangeEndKey) >= 0 {
 			break
 		}
-		// startKey in [ranges.StartKey, ranges.EndKey), ranges begin with [startKey, ranges.EndKey)
-		if ranges.StartKey.Cmp(startKey) <= 0 && ranges.EndKey.Cmp(startKey) > 0 {
-			keyRanges = keyRanges[i:]
-			keyRanges[0].StartKey = startKey
+
+		loc, err := s.GetRegionCache().LocateKey(bo, startKey)
+		if err != nil {
+			return taskStat, err
+		}
+
+		endKey := loc.EndKey
+		isLast := len(endKey) == 0 || (len(rangeEndKey) > 0 && bytes.Compare(endKey, rangeEndKey) >= 0)
+		// If it is the last region
+		if isLast {
+			endKey = rangeEndKey
+		}
+
+		logutil.BgLogger().Info("send prepare flashback request", zap.Uint64("region_id", loc.Region.GetID()),
+			zap.String("start_key", hex.EncodeToString(startKey)), zap.String("end_key", hex.EncodeToString(endKey)))
+
+		req := tikvrpc.NewRequest(tikvrpc.CmdPrepareFlashbackToVersion, &kvrpcpb.PrepareFlashbackToVersionRequest{
+			StartKey: startKey,
+			EndKey:   endKey,
+		})
+
+		resp, err := s.SendReq(bo, req, loc.Region, flashbackTimeout)
+		if err != nil {
+			return taskStat, err
+		}
+		regionErr, err := resp.GetRegionError()
+		if err != nil {
+			return taskStat, err
+		}
+		if regionErr != nil {
+			return taskStat, errors.Errorf(regionErr.String())
+		}
+		if resp.Resp == nil {
+			return taskStat, errors.Errorf("prepare flashback missing resp body")
+		}
+		prepareFlashbackToVersionResp := resp.Resp.(*kvrpcpb.PrepareFlashbackToVersionResponse)
+		if err := prepareFlashbackToVersionResp.GetError(); err != "" {
+			return taskStat, errors.Errorf(err)
+		}
+		taskStat.CompletedRegions++
+		if isLast {
 			break
 		}
+		bo = tikv.NewBackoffer(ctx, flashbackMaxBackoff)
+		startKey = endKey
 	}
+	return taskStat, nil
+}
 
-	return keyRanges, nil
+func prepareFlashbackToVersion(
```

Comment:
It seems that this function code is the same as `flashbackToVersion`. Is it duplicated?